llvm/include/llvm/Target/GlobalISel/Combine.td

```diff
@@ -1884,6 +1881,28 @@ def cast_combines: GICombineGroup<[
   buildvector_of_truncate
 ]>;
 
+def visit_icmp : GICombineRule<
```

Comment:
We can't have this kind of generic combine or we'll end up just copying DAGCombine/InstCombine internally. At the very least it needs to be renamed to be more specific otherwise people will keep trying to add miscellaneous things into this particular combine.

Reply:
I am open to renaming it, but it has a few optimizations that are hard to put into separate patterns. Furthermore, for some optimizations order matters. We have no good mechanism to control the order of combines. I don't want to see a fleet of combines that wip_match_opcode G_ICMP.

Reply:
Combining compares is a good and hard sport. Maybe different rules apply.

CombinerHelper::visitICmp first tries to constant fold the compare, if that fails it tries to canonicalize the constant to right. It is hard to put that into two separate patterns, which are not wip_match_opcode G_ICMP.

I have no interest to ever add a visitMULH. Adding a visitFCmp is a different story.

Reply:
We cannot express constant build vectors in MIR currently and visitFCmp would rely on `computeKnownFPClass`, which is more work.